File: src/Chapter4/EndProduct.java
package Chapter4;

/*                                                                            *
 * @author Gianluca Capone & Davide Sgobba                                    *
 *                                                                            */

/* This class contains all variables that define the computer product and the *
 * methods that operate over these variables                                  */
class EndProduct {
    
    // VARIABLES
    double cheap;
    /* Cheapness of the computer product (Z-CH_f,t)                           */
    double mod;
    /* Merit of design of the computer product (M_f,t)                        */
    double modForCust;
    /* Merit of design as perceived by customers of user class h (M_f,h,t)    */
    double perf;
    /* Performance of the computer product (Z-PE_f,t)                         */
    double productionCost;
    /* Computer production cost (C-CO_f,t)                                    */
    double u;
    /* Propensity of computer to be sold to customers (u_f,h,t)               */
    double U;
    /* Probability of computer to be sold to customers (U_f,h,t)              */

    //TECHNICAL VARIABLES & OBJECTS
    ComputerFirm Firm;
    /* Access to the firm                                                     */
    
    // CONSTRUCTOR
    EndProduct(ComputerFirm FIRM) {
        cheap          = 0;
        perf           = 0;
        mod            = 0;
        modForCust     = 0;
        productionCost = 0;
        u              = 0;
        U              = 0;
        Firm           = FIRM;
    }

    /* This method computes the mod of the computer based on the level of the *
     * mod of the component and the system element                            */
    void calcMod() {
        if (Firm.SysteM.mod > Firm.ComputerMarket.limitSysMod[Firm.tId]) {
            /* Equation 1. See also footnote 14 about technological limits    *
             * for the mod of the system element                              */
            mod = Firm.ComputerMarket.phi * Math.pow(
                    Firm.ComputerMarket.tau
                    * Math.pow(Firm.Component.mod, -Firm.ComputerMarket.rho)
                    + (1 - Firm.ComputerMarket.tau)
                    * Math.pow(Firm.ComputerMarket.limitSysMod[Firm.tId], -Firm.ComputerMarket.rho),
                      -(1 / Firm.ComputerMarket.rho));
        }
        else {
            /* Equation 1                                                     */
            mod = Firm.ComputerMarket.phi * Math.pow(
                    Firm.ComputerMarket.tau
                    * Math.pow(Firm.Component.mod, -Firm.ComputerMarket.rho)
                    + (1 - Firm.ComputerMarket.tau)
                    * Math.pow(Firm.SysteM.mod, -Firm.ComputerMarket.rho),
                      -(1 / Firm.ComputerMarket.rho));
        }
    }

    /* This method computes cheapness and performance of the computer, given  *
     * its mod                                                                */
    void calcCheapPerf() {
        cheap = mod * Math.cos(Firm.ComputerMarket.theta);
        perf  = mod * Math.sin(Firm.ComputerMarket.theta);
    }
    
    /* This method computes the production cost of the computer               */
    void calcCost() {
        /* Equation 9                                                         */
        productionCost = Firm.price / (1 + Firm.ComputerMarket.markup);
    }

    /* This method computes the mod as perceived by customers in the specific *
     * user class and the corresponding propensity to sell to these customers */
    void calcPropToSell() {
        /* Equation 3                                                         */
        modForCust = Math.pow(cheap, Firm.ComputerMarket.gamma)
                   * Math.pow(perf, 1 - Firm.ComputerMarket.gamma);
        /* Equation 2                                                         */
        u          = Math.pow(modForCust, Firm.ComputerMarket.deltaMod)
                   * Math.pow(1 + Firm.share, Firm.ComputerMarket.deltaShare);
    }
    
    /* This method computes the probability to sell                           */
    void calcProbToSell(double SUMPTS) {
        /* Equation 4                                                         */
        U = u/SUMPTS;
    }
    
    /* This method is activated when exit occurs: it resets the most relevant *
     * variables at the product level                                         */
    void exitComputer() {
        cheap          = 0;
        perf           = 0;
        mod            = 0;
        modForCust     = 0;
        productionCost = 0;
        u              = 0;
        U              = 0;
    }
    
}

File: src_py/Chapter4/end_product.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
EndProduct模块 - EndProduct类的Python实现
转换自Java版本的EndProduct.java
"""

import math

"""
@author Gianluca Capone & Davide Sgobba
Python转换

此类包含定义计算机产品的所有变量以及操作这些变量的方法
"""
class EndProduct:
    
    def __init__(self, firm):
        """
        构造函数
        
        Args:
            firm: 公司对象引用
        """
        # 变量
        self.cheap = 0.0           # 计算机产品的低成本性 (Z-CH_f,t)
        self.perf = 0.0            # 计算机产品的性能 (Z-PE_f,t)
        self.mod = 0.0             # 计算机产品的设计优点 (M_f,t)
        self.mod_for_cust = 0.0    # 用户类h的客户感知的设计优点 (M_f,h,t)
        self.production_cost = 0.0  # 计算机生产成本 (C-CO_f,t)
        self.u = 0.0               # 计算机销售给客户的倾向 (u_f,h,t)
        self.U = 0.0               # 计算机销售给客户的概率 (U_f,h,t)
        
        # 技术变量和对象
        self.firm = firm           # 访问公司对象
        
    def calc_mod(self):
        """
        根据组件和系统元素的mod水平计算计算机的mod
        """
        if self.firm.system.mod > self.firm.computer_market.limit_sys_mod[self.firm.t_id]:
            # 方程1，参见关于系统元素mod技术限制的脚注14
            self.mod = self.firm.computer_market.phi * math.pow(
                self.firm.computer_market.tau 
                * math.pow(self.firm.component.mod, -self.firm.computer_market.rho)
                + (1 - self.firm.computer_market.tau)
                * math.pow(self.firm.computer_market.limit_sys_mod[self.firm.t_id], -self.firm.computer_market.rho),
                -(1 / self.firm.computer_market.rho))
        else:
            # 方程1
            self.mod = self.firm.computer_market.phi * math.pow(
                self.firm.computer_market.tau
                * math.pow(self.firm.component.mod, -self.firm.computer_market.rho)
                + (1 - self.firm.computer_market.tau)
                * math.pow(self.firm.system.mod, -self.firm.computer_market.rho),
                -(1 / self.firm.computer_market.rho))
                
    def calc_cheap_perf(self):
        """
        根据计算机的mod计算其低成本性和性能
        """
        self.cheap = self.mod * math.cos(self.firm.computer_market.theta)
        self.perf = self.mod * math.sin(self.firm.computer_market.theta)
        
    def calc_cost(self):
        """
        计算计算机的生产成本
        """
        # 方程9
        self.production_cost = self.firm.price / (1 + self.firm.computer_market.markup)
        
    def calc_prop_to_sell(self):
        """
        计算特定用户类感知的mod和向这些客户销售的倾向
        """
        # 方程3
        self.mod_for_cust = (math.pow(self.cheap, self.firm.computer_market.gamma) 
                           * math.pow(self.perf, 1 - self.firm.computer_market.gamma))
        # 方程2
        self.u = (math.pow(self.mod_for_cust, self.firm.computer_market.delta_mod)
                * math.pow(1 + self.firm.share, self.firm.computer_market.delta_share))
                
    def calc_prob_to_sell(self, sum_pts):
        """
        计算销售概率
        
        Args:
            sum_pts: 所有倾向总和
        """
        # 方程4
        self.U = self.u / sum_pts
        
    def exit_computer(self):
        """
        退出时激活的方法：重置产品层面的最相关变量
        """
        self.cheap = 0.0
        self.perf = 0.0
        self.mod = 0.0
        self.mod_for_cust = 0.0
        self.production_cost = 0.0
        self.u = 0.0
        self.U = 0.0 